File: rdf4j-java/src/main/java/eu/neverblink/jelly/convert/rdf4j/Rdf4jEncoderConverter.java
package eu.neverblink.jelly.convert.rdf4j;

import eu.neverblink.jelly.core.NodeEncoder;
import eu.neverblink.jelly.core.ProtoEncoderConverter;
import eu.neverblink.jelly.core.RdfProtoSerializationError;
import eu.neverblink.jelly.core.utils.QuadDecoder;
import eu.neverblink.jelly.core.utils.TripleDecoder;
import org.eclipse.rdf4j.model.*;
import org.eclipse.rdf4j.model.vocabulary.XSD;

public class Rdf4jEncoderConverter
    implements ProtoEncoderConverter<Value>, TripleDecoder<Value, Statement>, QuadDecoder<Value, Statement> {

    @Override
    public void nodeToProto(NodeEncoder<Value> encoder, Value value) {
        if (value instanceof IRI iri) {
            encoder.makeIri(iri.stringValue());
        } else if (value instanceof BNode bNode) {
            encoder.makeBlankNode(bNode.getID());
        } else if (value instanceof Literal literal) {
            final var lex = literal.getLabel();
            final var lang = literal.getLanguage();
            if (lang.isPresent()) {
                encoder.makeLangLiteral(literal, lex, lang.get());
            } else {
                final var dt = literal.getDatatype();
                if (dt != XSD.STRING) {
                    encoder.makeDtLiteral(literal, lex, dt.stringValue());
                } else {
                    encoder.makeSimpleLiteral(lex);
                }
            }
        } else if (value instanceof Triple triple) {
            encoder.makeQuotedTriple(triple.getSubject(), triple.getPredicate(), triple.getObject());
        } else {
            throw new RdfProtoSerializationError("Cannot encode node: %s".formatted(value));
        }
    }

    @Override
    public void graphNodeToProto(NodeEncoder<Value> encoder, Value value) {
        if (value instanceof IRI iri) {
            encoder.makeIri(iri.stringValue());
        } else if (value instanceof BNode bNode) {
            encoder.makeBlankNode(bNode.getID());
        } else if (value instanceof Literal literal) {
            final var lex = literal.getLabel();
            final var lang = literal.getLanguage();
            if (lang.isPresent()) {
                encoder.makeLangLiteral(literal, lex, lang.get());
            } else {
                final var dt = literal.getDatatype();
                if (dt != XSD.STRING) {
                    encoder.makeDtLiteral(literal, lex, dt.stringValue());
                } else {
                    encoder.makeSimpleLiteral(lex);
                }
            }
        } else if (value == null) {
            encoder.makeDefaultGraph();
        } else {
            throw new RdfProtoSerializationError("Cannot encode graph node: %s".formatted(value));
        }
    }

    @Override
    public Value getQuadSubject(Statement statement) {
        return statement.getSubject();
    }

    @Override
    public Value getQuadPredicate(Statement statement) {
        return statement.getPredicate();
    }

    @Override
    public Value getQuadObject(Statement statement) {
        return statement.getObject();
    }

    @Override
    public Value getQuadGraph(Statement statement) {
        return statement.getContext();
    }

    @Override
    public Value getTripleSubject(Statement triple) {
        return triple.getSubject();
    }

    @Override
    public Value getTriplePredicate(Statement triple) {
        return triple.getPredicate();
    }

    @Override
    public Value getTripleObject(Statement triple) {
        return triple.getObject();
    }
}

File: rdf4j/src/main/scala/eu/ostrzyciel/jelly/convert/rdf4j/Rdf4jEncoderConverter.scala
package eu.ostrzyciel.jelly.convert.rdf4j

import eu.ostrzyciel.jelly.core.*
import eu.ostrzyciel.jelly.core.proto.v1.{GraphTerm, SpoTerm}
import org.eclipse.rdf4j.model.*
import org.eclipse.rdf4j.model.vocabulary.XSD

final class Rdf4jEncoderConverter extends ProtoEncoderConverter[Value, Statement, Statement]:

  override def getTstS(triple: Statement) = triple.getSubject
  override def getTstP(triple: Statement) = triple.getPredicate
  override def getTstO(triple: Statement) = triple.getObject

  override def getQstS(quad: Statement) = quad.getSubject
  override def getQstP(quad: Statement) = quad.getPredicate
  override def getQstO(quad: Statement) = quad.getObject
  override def getQstG(quad: Statement) = quad.getContext

  /** @inheritdoc */
  override def nodeToProto(encoder: NodeEncoder[Value], node: Value): SpoTerm = node match
    // URI/IRI
    case iri: IRI => encoder.makeIri(iri.stringValue)
    // Blank node
    case bnode: BNode => encoder.makeBlankNode(bnode.getID)
    // Literal
    case literal: Literal =>
      val lex = literal.getLabel
      val lang = literal.getLanguage
      if lang.isPresent then
        encoder.makeLangLiteral(literal, lex, lang.get)
      else
        val dt = literal.getDatatype
        if dt != XSD.STRING then
          encoder.makeDtLiteral(literal, lex, dt.stringValue)
        else
          encoder.makeSimpleLiteral(lex)
    case triple: Triple => encoder.makeQuotedTriple(
      nodeToProto(encoder, triple.getSubject),
      nodeToProto(encoder, triple.getPredicate),
      nodeToProto(encoder, triple.getObject),
    )
    case _ =>
      throw RdfProtoSerializationError(s"Cannot encode node: $node")

  /** @inheritdoc */
  override def graphNodeToProto(encoder: NodeEncoder[Value], node: Value): GraphTerm = node match
    // URI/IRI
    case iri: IRI => encoder.makeIri(iri.stringValue)
    // Blank node
    case bnode: BNode => encoder.makeBlankNode(bnode.getID)
    // Literal
    case literal: Literal =>
      val lex = literal.getLabel
      val lang = literal.getLanguage
      if lang.isPresent then
        encoder.makeLangLiteral(literal, lex, lang.get)
      else
        val dt = literal.getDatatype
        if dt != XSD.STRING then
          encoder.makeDtLiteral(literal, lex, dt.stringValue)
        else
          encoder.makeSimpleLiteral(lex)
    // Default graph
    case null => NodeEncoder.makeDefaultGraph
    case _ =>
      throw RdfProtoSerializationError(s"Cannot encode graph node: $node")
